Modularize Attitude Control

MaEtUgR opened the pull request:
**Describe problem solved by the proposed pull request**
Like discussed with @dagar separating core control calculations from organizational logic like task management with clear interfaces is desired for code readability, simple automatic unit testing, reusability (e.g. fixed wing attitude) and replaceability.

So I'm starting here with the example for the attitude control logic which I would know very well how to unit test because I did a lot of testing with hardcoded values before pushing it here: #8003 I had to get rid of all my test because there was no simple way to contribute them as well. My first goal is to enable them.

**Describe your preferred solution**
All attitude control calculations are separated into the `AttitudeControl` class which can then have its unit test set that allows for automatic evaluation. This class should also be more readable due to clear members and interfaces, is hence easier to extend and improve further and could also be reused for fixed wing attitude control.

MaEtUgR commented:
I made a quick and dirty test setup with [gtest](https://github.com/google/googletest) inspired by [DronecodeSDK](https://github.com/Dronecode/DronecodeSDK) testing and the first even so trivial corner case unit test I write fails: 
![trivial_test](https://user-images.githubusercontent.com/4668506/51803266-7d389180-2253-11e9-86a8-eb49f4eedffc.PNG)
I see high value in enabling such tests for the whole codebase. I know some tests are already possible but writing them is tedious and running them takes forever which results in people including me not using them. Can I join the discussion about how to improve that?

dagar commented:
I'd be interested in seeing the unit test integration. The one thing I really wanted to figure out is unit testing we can also run on the actual embedded hardware. So far I've failed with gtest (platform and c++ issues), but I think Catch2 might be workable. https://github.com/catchorg/Catch2

MaEtUgR commented:
@dagar I looked into Catch2 yesterday. For the basic tests it's more simple to include buuut on the long run I fear it might strike us back because it has far less customizability than gtest and also e.g. no built in mocking. It could end up with fast setup for embedded and a lot more work later.

MaEtUgR commented:
Devcall: Consent unit tests are necessary and valuable and should be integrated in the code base such that tests can be added with really low effort. Major concern we also need unit testing on hardware. I start the discussion about that with @dagar anyone who wants to join please write. I'll should probably create an issue for that whole topic.

MaEtUgR commented:
@dagar My private hacking branch is here: https://github.com/PX4/Firmware/compare/master...MaEtUgR:gtest but I can not yet encourage anyone to look at it otherwise I would have pushed it here. I found this interesting about running gtest on hardware: https://stackoverflow.com/questions/38427998/google-test-on-bare-metal-stm32-mcu

MaEtUgR commented:
@dagar ok, now that part is done and I'll follow up with the unit test pr.

dagar commented on src/modules/mc_att_control/AttitudeControl/AttitudeControl.cpp:
```diff
@@ -0,0 +1,111 @@
+	Vector3f e_z_d = qd.dcm_z();
+	Quatf qd_red(e_z, e_z_d);
+
+	if (abs(qd_red(1)) > (1.f - 1e-5f) || abs(qd_red(2)) > (1.f - 1e-5f)) {
```
```suggestion
	if (fabs(qd_red(1)) > (1.f - 1e-5f) || fabs(qd_red(2)) > (1.f - 1e-5f)) {
```

MaEtUgR replied:
Uuh, that's a good catch 👍

dagar commented on src/modules/mc_att_control/AttitudeControl/AttitudeControl.cpp:
```diff
@@ -0,0 +1,111 @@
+	qd.normalize();
+
+	// calculate reduced desired attitude neglecting vehicle's yaw to prioritize roll and pitch
+	Vector3f e_z = q.dcm_z();
```
For readability and general understanding I personally find it helpful to make these const if they aren't changed below in the function.

MaEtUgR replied:
Totally, if they aren't I forgot.

dagar commented on src/modules/mc_att_control/AttitudeControl/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+	AttitudeControl.cpp
+)
+target_include_directories(AttitudeControl
+PUBLIC
```
```suggestion
	PUBLIC
```

dagar commented on src/modules/mc_att_control/AttitudeControl/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+)
+target_include_directories(AttitudeControl
+PUBLIC
+	${CMAKE_CURRENT_SOURCE_DIR}
```
Actually, is this right? Shouldn't it be `INTERFACE`?

MaEtUgR replied:
I'll look it up, thanks.

MaEtUgR replied:
Today I learned from [stackoverflow](https://stackoverflow.com/questions/26243169/cmake-target-include-directories-meaning-of-scope) (the [cmake doc](https://cmake.org/cmake/help/v3.0/command/target_include_directories.html) is not as clear):
`PRIVATE`: only include for compiling the target
`INTERFACE`: only include for compiling dependencies
`PUBLIC`: include in both cases
With `INTERFACE` it fails to include the header in `AttitudeControl.cpp` with `#include <AttitudeControl.hpp>` but works with the explicit path `#include "AttitudeControl.hpp"`. I thought the first method combined with `PUBLIC` is preferable for the case you change the file structure of the target library, does anything speak against it?

MaEtUgR replied:
After a bit more research I consider this assumption to be confirmed.

dagar replied:
Makes sense. In other parts of the code base this hasn't come up because we have the top level `src` and `src/lib` included. The way you've done it is better and gives us more control.

dagar commented on src/modules/mc_att_control/mc_att_control_main.cpp:
```diff
@@ -542,75 +541,15 @@ MulticopterAttitudeControl::control_attitude()
-		qd_red = qd;
+	if ((_v_control_mode.flag_control_velocity_enabled || _v_control_mode.flag_control_auto_enabled) &&
+		!_v_control_mode.flag_control_manual_enabled) {
+		_attitude_control.setRateLimit(_auto_rate_max);
```
Don't we only need to update the rate limit on control mode change?

MaEtUgR replied:
In principle yes, I now first only made gains only update when they change but the plan clearly is to improve it further.

MaEtUgR replied:
I addressed this comment in https://github.com/PX4/Firmware/pull/11308/commits/71b9d69c123b4cce8304343a5f58c1b291c7542c

I think this extra logic should soon be removed completely. The contraints get part of the controller interface and are received together with the setpoint for example like here: https://github.com/PX4/Firmware/blob/8947cb5841acae5de58716b8933525851db975a9/src/modules/mc_pos_control/mc_pos_control_main.cpp#L742-L749 the generator of the attitude setpoints e.g. the position controller should decide what limits go well with the setpoints it publishes.

MaEtUgR commented:
@dagar 
- rebased on master without conflict ([diff](https://github.com/PX4/Firmware/compare/2bcfa4739b47bb6724c93fe5a8cf2ab9ceff5f53..16cb44bffa582afe1267bce517c8a49373d09225))
- refactored some parameter handling for gains and rate limits that catched my eyes in https://github.com/PX4/Firmware/pull/11308/commits/4556184bd8efc829f03429febb79d55cd3664956
- finally addressed your last open review comment in https://github.com/PX4/Firmware/pull/11308/commits/71b9d69c123b4cce8304343a5f58c1b291c7542c

MaEtUgR commented:
Is this good to go?

MaEtUgR commented:
Devcall: No objections for merging even before the release since it's just refactoring to enable unit testing. But we need hardware tests to make sure there is no problem with CPU and Memory on the microcontroller. So we add it to the flight test teams list and if possible do a HITL test.

jorge789 commented:
Tested on Pixhawk v2 Cube; No issues were noted, good flight, in general, tested in Position, Altitude, Stabilized, Mission and RTL.

https://review.px4.io/plot_app?log=7dccc216-aa4b-4d6e-9793-a2172d9461c1

Junkim3DR commented:
Testes on Pixhawk 4 v5; no issues were noted, good flight in general.
https://review.px4.io/plot_app?log=cd6a998a-fddf-489a-ae10-c5c01090a1df

MaEtUgR commented:
Thanks team for all the testing!! Mergeable @dagar ?

bkueng commented on src/modules/mc_att_control/mc_att_control.hpp:
```diff
@@ -58,6 +58,8 @@
 #include <uORB/topics/vehicle_land_detected.h>
 #include <uORB/topics/landing_gear.h>
 
+#include <AttitudeControl.hpp>
```
Within a module, I somewhat prefer relative includes, which makes it easier to see where the file is and reduces potential for conflicts:
```suggestion
#include "AttitudeControl/AttitudeControl.hpp"
```

MaEtUgR replied:
Thanks for the quick reviewing! I totally see your point if the class is closely tied to the module and lives within it's CMake target. But the whole point of this pr is to modularize the controller into a reusable self-contained library (https://github.com/PX4/Firmware/pull/11308/files#diff-66b6bc5cf0e74d3ebb07ebb30d50ade2R34) that is included and linked by CMake via the separate `AttitudeControl` library target (https://github.com/PX4/Firmware/pull/11308/files#diff-59b5eaaeffc3071a322da5225c7e3706R48). This makes it file location independent as long as the subdirectory is recursively added (https://github.com/PX4/Firmware/pull/11308/files#diff-59b5eaaeffc3071a322da5225c7e3706R34). If the file location is a problem we can already move it to the libraries directory but I thought I leave it in the mc_att_control folder for now since this is the first step only refactoring the existing functionality. I wanted to separate this step from introduction of testing, extending and reusing to make overseeable prs which are easy to review. (Testing draft here: https://github.com/PX4/Firmware/pull/11573)

bresch commented on src/modules/mc_att_control/mc_att_control.hpp:
```diff
@@ -145,6 +147,10 @@ class MulticopterAttitudeControl : public ModuleBase<MulticopterAttitudeControl>
 	matrix::Vector3f pid_attenuations(float tpa_breakpoint, float tpa_rate);
 
+	/** lower yawspeed limit in auto modes because we expect yaw steps */
+	void adapt_auto_yaw_rate_limit();
```
That's okay for now, but this changes the closed-loop response.
Let's take the extreme case where the user tightly tuned the yaw P gain to its maximum and is just stable; changing the saturation in auto mode will lead to yaw oscillations or even instability.
I made some modifications in the FlightTask to limit the rate of change of yaw setpoint such that we don't need to do that nasty thing here, I'll try to upstream that really soon.

MaEtUgR replied:
Good point, I totally agree. We definitely need to change that.
- This is refactoring only
- I made the function such that the logic only gets called in the event mode or parameter changes and not in every loop iteration, see review comment https://github.com/PX4/Firmware/pull/11308/files#discussion_r261836721
- I changed it to only switch the yaw rate limit because the rest was just confusing overhead
- The attitude controller currently can produce wobbles in the yaw movement when commanding big steps depending on the tuning gins which currently change the yaw weight (https://github.com/PX4/Firmware/pull/11308/files#diff-c7e26e8b748c35ed8718b27314fa807bR51). This stayed as it is because I'd like to have testing in place first (https://github.com/PX4/Firmware/pull/11573) and then change it according to my findings while experimenting with the prototyping script https://github.com/Auterion/Flight_Control_Prototyping_Scripts/tree/2eef448fb589a1c694e0717b5835644acf8d69dd/quaternion_attitude_control

MaEtUgR commented:
Thanks for the reviews. As I wrote this is the refactoring step only. I'll follow up with testing and changes to the behavior as soon as this is merged and I consider it safe now.
